dev/io.openliberty.data.internal_fat_jpa/test-applications/DataJPATestApp/src/test/jakarta/data/jpa/web/hibernate/DataJPAHibernateServlet.java

```diff
@@ -136,4 +146,101 @@ public void testCountQuery_WHERE_SELECT_ORDER() {
         assertEquals(10, page1.totalElements());
     }
 
+    /**
+     * Reproduces an issue where a previously detached entity is merged to the
+     * persistence context in order to make an update, but Hibernate never
+     * writes the update to the database, even when the transaction commits.
+     * This can also be used to reproduce a similar error where instead of a
+     * detached entity, the entity is fetched during the same transaction and
+     * then updated, with Hibernate similarly never writing the update to the
+     * database, even when the transaction commits. However, if we instead
+     * create a new instance of the entity and merge it to the persistence
+     * context, then Hibernate does write the update to the database.
+     */
+    @Test
+    public void testMergeDetachedEntity() throws Exception {
+        EntityManagerFactory emf = InitialContext
+                        .doLookup("java:comp/env/persistence/HibernatePersistenceUnitRef");
+        UserTransaction tx = InitialContext
+                        .doLookup("java:comp/UserTransaction");
+
+        SimpleEntity entity = new SimpleEntity();
+        entity.id = 1;
+        entity.value = "new";
+
+        EntityManager em = null;
+
+        tx.begin();
+        try {
+            em = emf.createEntityManager();
+            em.setCacheRetrieveMode(CacheRetrieveMode.BYPASS);
+            assertEquals(true, em.isJoinedToTransaction());
+
+            em.persist(entity);
+            // em.flush(); // unnecessary due to commit
+        } finally {
+            if (tx.getStatus() == Status.STATUS_ACTIVE)
+                tx.commit();
+            else
+                tx.rollback();
+            em.clear();
+            em.close();
+        }
+
+        try {
+            em = emf.createEntityManager();
+            em.setCacheRetrieveMode(CacheRetrieveMode.BYPASS);
+
+            entity = em.find(SimpleEntity.class, 1);
+        } finally {
+            em.clear();
+            em.close();
+        }
+
+        assertEquals("new", entity.value);
+
+        entity = new SimpleEntity(); // workaround
+        entity.id = 1; // workaround
+        // TODO remove this line and 2 lines above once #33232 is fixed, or to reproduce
+        entity.value = "merged";
+
+        tx.begin();
+        try {
+            em = emf.createEntityManager();
+            em.setCacheRetrieveMode(CacheRetrieveMode.BYPASS);
+            assertEquals(true, em.isJoinedToTransaction());
+
+            SimpleEntity found = (SimpleEntity) em
+                            .createQuery("SELECT this FROM SimpleEntity WHERE id=?1")
+                            .setParameter(1, 1L)
+                            .setLockMode(LockModeType.PESSIMISTIC_WRITE)
+                            .getResultList()
+                            .get(0);
```

Comment:
```suggestion
            SimpleEntity found = em.find(SimpleEntity.class, 1, LockModeType.PESSIMISTIC_WRITE);
```
suggestion: simpler way to write this.

Reply:
Yes, that is simpler (and I also tried it that way) but wanted the test case to match what the code is currently doing.  The code is that way because we were going to look into the optimization of only returning the id rather than the whole entity.  Although at this point, we can probably just switch it to `find`.